auth0/resource_auth0_user.go

```diff
@@ -250,21 +250,54 @@ func expandUser(d *schema.ResourceData) (*management.User, error) {
 		Blocked:       Bool(d, "blocked"),
 	}
 
-	userMeta, err := JSON(d, "user_metadata")
+	userMetadata, err := expandMetadata(d, "user")
 	if err != nil {
 		return nil, err
 	}
-	user.UserMetadata = &userMeta
+	user.UserMetadata = &userMetadata
 
-	appMeta, err := JSON(d, "app_metadata")
+	appMetadata, err := expandMetadata(d, "app")
 	if err != nil {
 		return nil, err
 	}
-	user.AppMetadata = &appMeta
+	user.AppMetadata = &appMetadata
 
 	return user, nil
 }
 
+func expandMetadata(d *schema.ResourceData, metadataType string) (map[string]interface{}, error) {
+	if !d.HasChange(metadataType + "_metadata") {
+		return map[string]interface{}{}, nil
+	}
+
+	oldMetadata, newMetadata := d.GetChange(metadataType + "_metadata")
+	if oldMetadata == "" {
+		return JSON(d, metadataType+"_metadata")
+	}
+
+	if newMetadata == "" || newMetadata == nil {
+		return map[string]interface{}{}, nil
+	}
+
+	oldMap, err := structure.ExpandJsonFromString(oldMetadata.(string))
+	if err != nil {
+		return map[string]interface{}{}, err
+	}
+
+	newMap, err := structure.ExpandJsonFromString(newMetadata.(string))
+	if err != nil {
+		return map[string]interface{}{}, err
+	}
+
+	for key := range oldMap {
+		if _, ok := newMap[key]; !ok {
+			newMap[key] = nil
```

Comment:
We're applying the following explained logic in https://github.com/alexkappa/terraform-provider-auth0/issues/392#issuecomment-884198345, where in order to properly remove something we need to send it's value as null. 

This is because the metadata fields are merged instead of being replaced but the merge only occurs on the first level during a PATCH.